Add KeyVaultSettingsClient

heaths opened the pull request:
- [x] Add tests for `KeyVaultSettingsClient`
- [x] Get approval on overloads to set type from @KrzysztofCwalina and @tg-msft; should we still then even make the `KeyVaultSetting` public since it has a string value? /cc @vcolin7 
- [x] Re-record all tests for the Administration package for 7.4-preview.1. Re-recording tests for other packages is tracked in #31982

heaths committed:
Add KeyVaultSettingsClient

vcolin7 commented on sdk/keyvault/Azure.Security.KeyVault.Administration/api/Azure.Security.KeyVault.Administration.netstandard2.0.cs:
```diff
@@ -281,4 +283,41 @@ public partial class KeyVaultSelectiveKeyRestoreResult
+        public virtual System.Threading.Tasks.Task<Azure.Response<Azure.Security.KeyVault.Administration.KeyVaultSetting>> UpdateSettingAsync(string name, bool value, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
+    }
+    [System.Runtime.InteropServices.StructLayoutAttribute(System.Runtime.InteropServices.LayoutKind.Sequential)]
+    public readonly partial struct SettingType : System.IEquatable<Azure.Security.KeyVault.Administration.SettingType>
```
Should this be named `KeyVaultSettingType` if the public type is `KeyVaultSetting`?

heaths replied:
We have `KeyType` to `KeyVaultKey` so I think `SettingType` makes more sense...if we even keep it.

vcolin7 replied:
We actually have `KeyVaultRoleType` in Java, I can't recall the exact reasoning behind the name but I think it's because all other classes it relates to start with `KeyVaultRole`. Maybe it was because RoleType is something that can be found elsewhere for any library that offers RBAC functionality. In any case, I think for Java we will stick to precedent and .NET can probably do the same. I'll talk to Jonathan or Srikanta about this offline.

heaths committed:
Prototype type-specific getters

Similar to JsonElement's methods.

heaths committed:
Resolve some architect feedback

heaths committed:
Change KeyVaultSetting design to mimic Form Recognizer

heaths committed:
--amend

heaths committed:
More changes to KeyVaultSetting from architects

Also re-recorded all the tests with 7.4-preview.1 and added more tests, completing #31847.

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[Azure.Security.KeyVault.Administration](https://apiview.dev/Assemblies/Review/eff7ad59fc454f62835a2d2fe2f818f9)

heaths committed:
Update public API and samples

christothes commented on sdk/keyvault/Azure.Security.KeyVault.Administration/tests/KeyVaultSettingsClientLiveTests.cs:
```diff
@@ -0,0 +1,26 @@
+
+namespace Azure.Security.KeyVault.Administration.Tests
+{
+    public class KeyVaultSettingsClientLiveTests : SettingsTestBase
```
Do we need to test the other operations as well?

heaths replied:
They are in the sample that's also a recorded test. Didn't figure we need to test 2/3<sup>rds</sup> of the operations twice. Do you feel we should for some other reason?

heaths replied:
Or was it just that it wasn't obvious? I could add a comment pointing to the sample if that would help. KV tests already take long enough, I didn't want to compound it further unnecessarily. Sure, this is quick, but it all adds up.

heaths replied:
I added a comment. Let me know if you want more, though.

heaths commented on sdk/keyvault/Azure.Security.KeyVault.Administration/src/KeyVaultAdministrationModelFactory.cs:
```diff
@@ -120,5 +120,13 @@ public static KeyVaultRoleDefinition KeyVaultRoleDefinition(string id, string na
+        /// <param name="value"> The value of the pool setting. </param>
+        /// <param name="type"> The type specifier of the value. </param>
+        /// <returns> A new <see cref="Administration.KeyVaultSetting"/> instance for mocking. </returns>
+        public static KeyVaultSetting KeyVaultSetting(string name, string value, SettingType? type = null) =>
```
Given the factory is used for testing, at the very least, a setting type will expand over time. This was meant to provide an "escape" hatch if a different type is needed that isn't supported by swagger. The actual properties are not defined by swagger so, technically, the service could define a new property of a different type at any time. The setting type itself is `modelAsString: true`.

heaths replied:
@KrzysztofCwalina what do you think? Based on our conversation last week about hiding possible future types in the future, do you think the factory should allow for it? I feel it's a nice escape hatch, but also contrary to our conversation. Then again, that's what our factories generally allow in the name of testing.

KrzysztofCwalina replied:
I think we can add an overload once service supports other types, and for now just support bool.

heaths committed:
Use absolute URLs for links

heaths committed:
Regenerate after rebase on main

heaths committed:
Resolve PR feedback
